src/DNSC/UpdateCache.hs

```diff
@@ -0,0 +1,64 @@
+module DNSC.UpdateCache (
+  newCache,
+  ) where
+
+import Control.Concurrent (forkIO, threadDelay)
+import Control.Concurrent.Chan (newChan, readChan, writeChan)
+import Control.Monad (void, forever)
+import Data.IORef (newIORef, readIORef, writeIORef)
+import Data.Time (UTCTime, getCurrentTime)
+
+import Network.DNS (TTL, Domain, TYPE, CLASS, ResourceRecord)
+
+import DNSC.Cache (Cache, Key, Val, CRSet, Ranking)
+import qualified DNSC.Cache as Cache
+
+data Update
+  = I Key TTL CRSet Ranking
+  | E
+  deriving Show
+
+runUpdate :: UTCTime -> Update -> Cache -> Maybe Cache
+runUpdate t u = case u of
+  I k ttl crs rank -> Cache.insert t k ttl crs rank
+  E                -> Cache.expires t
+
+type Lookup = Domain -> TYPE -> CLASS -> IO (Maybe ([ResourceRecord], Ranking))
+type Insert = Key -> TTL -> CRSet -> Ranking -> IO ()
+type Dump = IO [(Key, (UTCTime, Val))]
+
+newCache :: (String -> IO ()) -> IO (Lookup, Insert, Dump)
+newCache putLog = do
+  let putLn = putLog . (++ "\n")
+  cacheRef <- newIORef Cache.empty
+  updateQ <- newChan
+
+  let update1 = do   -- step of single update theard
+        (ts, u) <- readChan updateQ
+        cache <- readIORef cacheRef
+        let updateRef c = do
+              writeIORef cacheRef c
+              case u of
+                I {}  ->  return ()
+                E     ->  putLn $ show ts ++ ": some records expired: size = " ++ show (Cache.size c)
+        maybe (pure ()) updateRef $ runUpdate ts u cache
+  void $ forkIO $ forever update1
```

Comment:
Who kills this thread?

Reply:
Since there is no termination process, this thread will not be killed by anyone.

Reply:
The termination process should be defined in the future, I think.